quickcomm/models.py

```diff
@@ -18,6 +24,145 @@
 # We used UUIDs for pkeys to be more secure.
 # TODO what are the constraints on fields being null?
 
+class HostAuthenticator(models.Model):
+    """A host authenticator is a username and password that can be used to
+    authenticate to a host."""
+
+    # Note: passwords are stored in plain-text. They are API keys, so they are
+    # not sensitive. If we want to store them in a more secure way, we can
+    # hash them, but for now, we will leave them as is.
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    username = models.CharField(max_length=100, help_text="The username to use to authenticate to the host.", verbose_name="Username")
+    password = models.CharField(max_length=100, help_text="The password to use to authenticate to the host.", verbose_name="Password")
+    nickname = models.CharField(max_length=100, help_text="The nickname of the host authenticator. This is only used for display purposes.", verbose_name="Nickname", null=True, blank=True)
+
+
+    @property
+    def nickname_or_username(self):
+        if self.nickname:
+            return self.nickname
+        return self.username
+
+    @property
+    def base64string(self):
+        return base64.b64encode(f"{self.username}:{self.password}".encode('ascii')).decode('ascii')
+
+    def check_password(self, password):
+        return self.password == password
+
+    @property
+    def is_authenticated(self):
+        return True
+
+    def __str__(self):
+        return f"{self.username}"
+
+class Host(models.Model):
+    """A host is a remote server that hosts authors."""
+
+    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    url = models.URLField(validators=[URLValidator], help_text="The URL of the host API. This must be in proper form (e.g. https://example.com/).", verbose_name="Host URL")
+
+    username_password_base64 = models.CharField(max_length=100, help_text="The username and password to use to authenticate to the host. This is a base64 encoded string of the form 'username:password'.", verbose_name="Username and Password", null=True, blank=True)
+
+    last_successful_ping = models.DateTimeField(null=True, blank=True)
+    last_ping = models.DateTimeField(null=True, blank=True)
+    last_ping_result = models.BooleanField(null=True, blank=True)
+
+    nickname = models.CharField(max_length=100, help_text="The nickname of the host. This is only used for display purposes.", verbose_name="Nickname", null=True, blank=True)
+
+    @property
+    def nickname_or_url(self):
+        if self.nickname:
+            return self.nickname
+        return self.url
+
+    def sync(self):
+        """Syncs the host with the remote API."""
+        self.sync_authors()
+
+    def sync_authors(self, admin=None, request=None):
```

Comment:
should this be a private method? If not then can we get rid of the 'sync' function above?